main.go

```diff
@@ -111,123 +151,134 @@ func main() {
-				errorMessage := fmt.Sprintf("RequestId: %s A platform error caused a shutdown", lastRequestId)
-				batch.AddTelemetry(lastRequestId, []byte(errorMessage))
+			if event.EventType == api.Shutdown {
+				if event.ShutdownReason == api.Timeout && lastRequestId != "" {
+					// Synthesize the timeout error message that the platform produces, and LLC parses
+					timestamp := eventStart.UTC()
+					timeoutSecs := eventStart.Sub(lastEventStart).Seconds()
+					timeoutMessage := fmt.Sprintf(
+						"%s %s Task timed out after %.2f seconds",
+						timestamp.Format(time.RFC3339),
+						lastRequestId,
+						timeoutSecs,
+					)
+					batch.AddTelemetry(lastRequestId, []byte(timeoutMessage))
+				} else if event.ShutdownReason == api.Failure && lastRequestId != "" {
+					// Synthesize a generic platform error. Probably an OOM, though it could be any runtime crash.
+					errorMessage := fmt.Sprintf("RequestId: %s A platform error caused a shutdown", lastRequestId)
+					batch.AddTelemetry(lastRequestId, []byte(errorMessage))
+				}
+
+				pollLogServer(logServer, batch)
+				finalHarvest := batch.Close()
+				shipHarvest(ctx, finalHarvest, telemetryClient, invokedFunctionARN)
+				return eventCounter
 			}
 
-			break
-		}
+			invokedFunctionARN = event.InvokedFunctionARN
+			lastRequestId = event.RequestID
 
-		invokedFunctionARN = event.InvokedFunctionARN
-		lastRequestId = event.RequestID
-		// Create an invocation record to hold telemetry
-		batch.AddInvocation(lastRequestId, eventStart)
+			// Create an invocation record to hold telemetry
+			batch.AddInvocation(lastRequestId, eventStart)
 
-		// Await agent telemetry. This may time out, so we race the timeout against the telemetry channel
-		// timeoutInstant is when the invocation will time out
-		timeoutInstant := time.Unix(0, event.DeadlineMs*int64(time.Millisecond))
+			// Await agent telemetry. This may time out
+			// timeoutInstant is when the invocation will time out
+			timeoutInstant := time.Unix(0, event.DeadlineMs*int64(time.Millisecond))
 
-		// Set the timeout timer for a smidge before the actual timeout; we can recover from early.
-		timeoutWatchBegins := time.Millisecond * 100
-		timeout := time.NewTimer(timeoutInstant.Sub(time.Now()) - timeoutWatchBegins)
-		select {
-		case telemetryBytes := <-telemetryChan:
-			// We received telemetry
-			util.Debugf("Agent telemetry bytes: %s", base64.URLEncoding.EncodeToString(telemetryBytes))
-			inv := batch.AddTelemetry(lastRequestId, telemetryBytes)
-			if inv == nil {
-				util.Logf("Failed to add telemetry for request %v", lastRequestId)
-			}
-			// Tear down the timer
-			if !timeout.Stop() {
-				<-timeout.C
-			}
+			// Set the timeout timer for a smidge before the actual timeout;
+			// we can recover from early.
+			timeoutWatchBegins := time.Millisecond * 100
+			timeout := timeoutInstant.Sub(time.Now()) - timeoutWatchBegins
+			util.Debugf("Timeout in %v", timeout)
```

Comment:
That log line is going to be really chatty. Moreso that the debug level has been, historically. Let's not ship with this in place, if possible.